### PR TITLE
Anemoi circuit to run all rounds

### PR DESCRIPTION
**Pull request by vesselinux**
Addresses https://github.com/clearmatics/libsnark/issues/104 .

### CODE REVIEW

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp [hunk 1]
```diff
@@ -263,6 +258,47 @@ template<typename ppT> class anemoi_permutation_mds<ppT, 4>
...
+
+private:
+    // C round constants for all rounds
+    std::vector<std::vector<FieldT>> C_const_vec;
+    // D round constants for all rounds
+    std::vector<std::vector<FieldT>> D_const_vec;
+    // vector of round gadgets
+    std::vector<
+        anemoi_round_prime_field_gadget<ppT, NumStateColumns, parameters>>
+        Round;
```
Comment (dtebbs) [inline libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp]: `Rounds`?

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp [hunk 2]
```diff
@@ -263,6 +258,47 @@ template<typename ppT> class anemoi_permutation_mds<ppT, 4>
...
+    typename ppT,
+    size_t NumStateColumns,
+    class parameters = anemoi_parameters<libff::Fr<ppT>>>
+class anemoi_permutation_prime_field_gadget : public gadget<libff::Fr<ppT>>
+{
+    using FieldT = libff::Fr<ppT>;
+
+private:
+    // C round constants for all rounds
+    std::vector<std::vector<FieldT>> C_const_vec;
```
Comment (dtebbs) [inline libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp:274]: Do (a copy of) these need to be stored on the class?

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc [hunk 3]
```diff
@@ -673,6 +716,135 @@ std::array<std::array<libff::Fr<ppT>, 4>, 4> anemoi_permutation_mds<ppT, 4>::
...
+    // Number of columns can not be larger than rounds128 size
+    assert(NumStateColumns <= parameters::nrounds128.size());
+    // Number of columns can not be larger than rounds256 size
+    assert(NumStateColumns <= parameters::nrounds256.size());
+
+    // Get the number of rounds for the given Anemoi instance
+    // (i.e. given number of columns in the state). Note: currently
+    // using 256-bit security instance by default. TODO add support
+    // for 128-bit security e.g. by adding a Boolean flag b_sec_128 in
+    // the tamplate parameters.
```
Comment (dtebbs) [inline libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc]: "template"

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc [hunk 4]
```diff
@@ -464,68 +464,120 @@ pb_linear_combination_array<libff::Fr<ppT>> anemoi_vector_left_rotate_by_one(
...
+void anemoi_round_prime_field_gadget<ppT, NumStateColumns, parameters>::
+    anemoi_get_round_constants(
+        const size_t &iround, std::vector<FieldT> &C, std::vector<FieldT> &D)
+{
+    assert(iround < parameters::nrounds256[NumStateColumns - 1]);
+    assert(C.size() == 0);
+    assert(D.size() == 0);
+
+    if (NumStateColumns == 1) {
+        copy(
```
Comment (dtebbs) [inline libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc:477]: do we need copies of these values?

**Review by dtebbs — COMMENTED** [review]
Overall looks good.  Just a couple of points related to copying vs referencing the constants.

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp (hunk 1 above)
Comment (vesselinux) [reply libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp]: Fixed.

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp (hunk 2 above)
Comment (vesselinux) [reply libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp:274]: Indeed, those constants are stored in the corresponding parameters class for each curve e.g. https://github.com/clearmatics/libsnark/blob/104-anemoi-permutation-all-rounds/libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_parameters_bls12_381.tcc . In principle they could be extracted from the `parameters` template parameter of the `anemoi_permutation_prime_field_gadget gadget` (https://github.com/clearmatics/libsnark/blob/104-anemoi-permutation-all-rounds/libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp#L267) . The problem is that the `parameters` class contains 4 sets of constants -- one for each of the 4 instances of Anemoi (resp. with 1,2,3 or 4 columns in the state). Which one of the four sets should be used is determined by the `NumStateColumns` template parameter (https://github.com/clearmatics/libsnark/blob/104-anemoi-permutation-all-rounds/libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp#L265) and the corresponding set is copied in the constructor of the gadget accordingly.

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc (hunk 3 above)
Comment (vesselinux) [reply libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc]: Outdated.

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc (hunk 4 above)
Comment (vesselinux) [reply libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc:477]: Is the concern here the same as in comment https://github.com/clearmatics/libsnark/pull/110#discussion_r1104829750 ? If so, please see my response https://github.com/clearmatics/libsnark/pull/110#discussion_r1126340930 .

**Pull request merged**